### DISCUSSION TITLE
Text file load with regex `read_text(regex)`

### DISCUSSION THREAD

**Discussion by pankajp** [duckdb/duckdb#13534] [Ideas]
I believe extracting data from text files/logs is a pretty common operation. While it can be done using the client language when using duckdb as a library, I believe aligning with duckdb's philosophy of user friendliness a text import function might be immensely useful.

For example for the following snippet of a log file
```
2006-02-08 22:20:02,165 [INFO] 192.168.0.1 fbloggs  new connection
 client is very old
2006-02-08 22:20:02,165 [WARN] 192.168.0.1 fbloggs  some useful warning
2006-02-08 22:20:02,165 [ERROR] 192.168.0.1 fbloggs  Protocol problem: connection reset
```

We can load important events from the logs as follows
```sql

CREATE TYPE LogLevel AS ENUM ('INFO', 'WARN', 'ERROR');

SELECT *
FROM read_text('connections.log',
    regex = '([0-9]{4}-[0-9]{2}-[0-9]{2} ..:..:..,...) \[(WARN|ERROR)] ([^ ]*) (.*)',
    columns = {
        'timestamp': 'TIMESTAMP',
        'level': 'LogLevel',
        'ipaddr': 'VARCHAR',
        'message': 'VARCHAR',
    });
```

The implementation could be to scan the file for regexp matches and extract and append the rows as directed from columns. If columns/types are not specified the type could be varchar (or implement csv like auto detection) and names could come from regex group index/name.
All unmatched text/lines would be dropped.

**Comment by rustyconover** [comment]
Hi @pankajp you can do this by filtering the input to read_text() via the shellfs extension here:

https://community-extensions.duckdb.org/extensions/shellfs.html

You could put a program that reads the input before DuckDB receives it, filters and reformats the input and DuckDB just works as normal.

**Reply by pankajp** [reply]
Thanks @rustyconover , I didn't know about shellfs, and while it is very powerful indeed, it is also something I'd like to disable when allowing others to execute user defined sql queries on my data.
I see there already is a read_text() function which reads the whole file as a single row, so maybe we need a different name.
Actually I thought I could just write a custom table function for use if this feature isn't implemented in duckdb, but then I realized duckdb doesn't expose writing a table function in the python api (maybe I don't know where to look?), though it is present in the C api.

xref: https://github.com/duckdb/duckdb/discussions/4797#discussioncomment-5765962

**Comment by lakemove** [comment]
maybe could use #16975 

```
from regexp_extract_table(log_content, '(?P<timestamp>[0-9]{4}-[0-9]{2}-[0-9]{2} ..:..:..,...) \[(?P<level>(WARN|ERROR))] (?P<ip>[^ ]*) (?P<msg>.*)')
```
